c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3676,6 +3681,133 @@ bool response_started(JSObject *self) {
 }
 } // namespace FetchEvent
 
+using std::chrono::system_clock;
+using TimerArgumentsVector = std::vector<JS::Heap<JS::Value>>;
+
+class Timer {
+public:
+  uint32_t id;
+  JS::Heap<JSObject *> callback;
+  TimerArgumentsVector arguments;
+  uint32_t delay;
+  system_clock::time_point deadline;
+  bool repeat;
+
+  Timer(uint32_t id, HandleObject callback, uint32_t delay, system_clock::time_point deadline,
+        JS::HandleValueVector args, bool repeat)
+      : id(id), callback(callback), delay(delay), deadline(deadline), repeat(repeat) {
+    deadline = system_clock::now() + system_clock::duration(delay * 1000);
+    arguments.reserve(args.length());
+    for (auto arg : args) {
+      arguments.push_back(JS::Heap(arg));
+    }
+  }
+
+  friend bool operator<(Timer const &left, Timer const &right) {
+    return left.deadline < right.deadline;
+  }
+
+  void trace(JSTracer *trc) {
+    JS::TraceEdge(trc, &callback, "Timer callback");
+    for (auto &arg : arguments) {
+      JS::TraceEdge(trc, &arg, "Timer callback arguments");
+    }
+  }
+};
+
+class ScheduledTimers {
+public:
+  Timer *first() { return *timers.begin(); }
```

Comment:
This will read arbitrary memory if the list is empty. Is it only called in cases where you know the list is not empty?

Reply:
It is only called when we know there is an item in the list, however I have also added an assert to this function and also changed it to use `.front()` instead:
```cpp
Timer *first() {
    MOZ_RELEASE_ASSERT(timers.size() > 0);
    return timers.front();
  }
```

Reply:
I've changed this once more now, it now checks to see if the list is empty and if so, it returns nullptr, otherwise it returns the front of the list